modules/mono/csharp_script.cpp

```diff
@@ -2419,166 +1966,66 @@ bool CSharpScript::_update_exports(PlaceHolderScriptInstance *p_instance_to_upda
 	if (exports_invalidated)
 #endif
 	{
-		GD_MONO_SCOPE_THREAD_ATTACH;
+#ifdef TOOLS_ENABLED
+		exports_invalidated = false;
+#endif
 
 		changed = true;
 
 		member_info.clear();
 
+		if (GDMonoCache::godot_api_cache_updated) {
+			GDMonoCache::managed_callbacks.ScriptManagerBridge_GetPropertyInfoList(this,
+					[](CSharpScript *p_script, const String *p_current_class_name, GDMonoCache::godotsharp_property_info *p_props, int32_t p_count) {
 #ifdef TOOLS_ENABLED
-		MonoObject *tmp_object = nullptr;
-		Object *tmp_native = nullptr;
-		uint32_t tmp_pinned_gchandle = 0;
-
-		if (is_editor) {
-			exports_invalidated = false;
-
-			exported_members_cache.clear();
-			exported_members_defval_cache.clear();
```

Comment:
It seems you accidentally removed this so the members are not cleared and when I rebuild a project the properties are duplicated.

Also, for some reason the exported members seem to be in reverse order:
![image](https://user-images.githubusercontent.com/3903059/183537756-4a9dc612-48db-4bde-94b8-f08d312aba5f.png)

But I can take a look at that if you want, after merging the PR.

Reply:
I don't know if this is a regression from #63712 or a mistake of mine when trying to rebase it. Does this happen in master? I was careful when rebasing this, but that PR was hard to grasp (I don't understand what it does), so I may have made a mistake.

Reply:
If I remember correctly before #63712 properties in C# scripts were all in the same category and that PR separated them in a category for each class in the hierarchy (also, it fixed an issue where GDScript properties showed up duplicated in the inspector but I don't think that was affecting C# scripts).

At some point #63712 accidentally reversed but the author eventually fixed it, I tested it to make sure it didn't introduce a regression. But I'll check in master again just to confirm.

Reply:
I've compiled master as of commit 7355dfb50 and it works correctly.

![image](https://user-images.githubusercontent.com/3903059/183554500-11354dc2-2fa1-405e-a170-9f26e0eb0363.png)

Reply:
Re-adding the two `clear` lines fixed the duplication for me. It seems the reverse order was happening because the legacy version was iterating the properties and fields in reverse order, unlike the new source generators.

Reply:
It's still showing duplicates for me as of 1c137e101, here's the inspector for a three level hierarchy (the longer the hierarchy the more duplicates):

![image](https://user-images.githubusercontent.com/3903059/183706426-7cfc5f28-031a-4c5f-bbf8-00caddd0ed16.png)

Reply:
I think I fixed it, but I'm a bit confused. I fixed it by making the script only add the properties from its class, without the ones from the base classes. However, the version from #63712 also added the ones from the base classes, so I don't know why that one didn't have this problem.

Reply:
I think it may have something to do with `_update_exports_values` calling `base_script->_update_exports_values`. The method is called in `_update_exports` to populate the `PlaceHolderScriptInstance` which is what is used when the script is not a `[Tool]` so I guess it gets populated with the base script properties too? So I tested making the scripts `[Tool]` and indeed now they only show the declared properties.

Reply:
> I think it may have something to do with `_update_exports_values` calling `base_script->_update_exports_values`.

Ah, yes! That code has no effect on master because `base_cache` (`base_script` now) was never implemented. Now it's implemented in this PR, hence why the base class properties were being added again. I went with the right fix then.

Reply:
I think you might've misunderstood me, right now non-`[Tool]` scripts show the properties correctly in the inspector but `[Tool]` scripts only show the properties declared in the script and not the inherited properties.

Reply:
Should be fixed now. `get_script_property_list` and `get_property_list` had to include inherited properties. This works on `master` because in that old version each script stores all properties, including the inherited ones.